src/tools/illink/src/linker/Linker/TypeMapInfo.cs

```diff
@@ -329,22 +336,26 @@ void AnnotateMethods (MethodDefinition @base, MethodDefinition @override, Interf
 		/// <param name="implOfInterface">
 		/// The InterfaceImplementation on <paramref name="type"/> that points to the DeclaringType of <paramref name="interfaceMethod"/>.
 		/// </param>
-		void FindAndAddDefaultInterfaceImplementations (TypeDefinition typeThatImplementsInterface, TypeDefinition typeThatMayHaveDIM, MethodDefinition interfaceMethodToBeImplemented, InterfaceImplementation originalInterfaceImpl)
+		void FindAndAddDefaultInterfaceImplementations (RuntimeInterfaceImplementation originalInterfaceImpl, MethodReference _, MethodDefinition interfaceMethodDef)
 		{
 			// Go over all interfaces, trying to find a method that is an explicit MethodImpl of the
 			// interface method in question.
+			if (!_runtimeInterfaceImpls.TryGetValue (originalInterfaceImpl.Implementor, out var runtimeIfaces))
+				return;
 
-			foreach (var interfaceImpl in typeThatMayHaveDIM.Interfaces) {
-				var potentialImplInterface = context.TryResolve (interfaceImpl.InterfaceType);
-				if (potentialImplInterface == null)
+			foreach (var interfaceImpl in runtimeIfaces) {
+				var potentialImplInterface = interfaceImpl.InterfaceTypeDefinition;
+				if (potentialImplInterface is null)
 					continue;
 
+				// We go by MethodDefinition here and may miss the correct implementation by exiting early after finding one.
+				// To correct for this, we'll look for all possible implementations and add them all if the inflatedIntefaceMethodReference is a genericInstance
+				// Otherwise, we're good to stop looking higher in the hierarchy after the first. Note we still need to look at other branches of the hierarchy to find any diamond cases.
 				bool foundImpl = false;
-
 				foreach (var potentialImplMethod in potentialImplInterface.Methods) {
-					if (potentialImplMethod == interfaceMethodToBeImplemented &&
+					if (potentialImplMethod == interfaceMethodDef &&
 						!potentialImplMethod.IsAbstract) {
-						AddDefaultInterfaceImplementation (interfaceMethodToBeImplemented, new (typeThatImplementsInterface, originalInterfaceImpl, interfaceMethodToBeImplemented.DeclaringType, context), potentialImplMethod);
+						AddDefaultInterfaceImplementation (interfaceMethodDef, originalInterfaceImpl, potentialImplMethod);
 						foundImpl = true;
 						break;
```

Comment:
In this case, the method we're looking at is the interface method we're trying to find an override for. Since it is the base method and provides an implementation, we can stop looking at the methods on this interface and move on to the next.